src/vm.rs

```diff
@@ -649,43 +648,55 @@ impl<T: GuestStagePageTable> Vm<T, VmStateFinalized> {
             };
 
             // TODO: Update internal counter state on success.
-            result.map_err(|e| EcallError::Sbi(e))
+            result
+                .map(|n| {
+                    active_cpu.update_started_counters(counter_index_base, counter_index_mask);
+                    n
+                })
+                .map_err(|e| EcallError::Sbi(e))
         }
 
         #[allow(clippy::redundant_closure)]
         fn stop_counters(
             counter_index_base: u64,
             counter_index_mask: u64,
             stop_flags: PmuCounterStopFlags,
-            _active_cpu: &mut VmCpu,
+            active_cpu: &mut VmCpu,
         ) -> EcallResult<u64> {
             use sbi::api::pmu;
-            let pmu_info = PmuInfo::get();
-            let result = if pmu_info.get_counter_info(counter_index_base).is_some() {
+            let result = if active_cpu.is_counter_stopabble(counter_index_base) {
                 pmu::stop_counters(counter_index_base, counter_index_mask, stop_flags)
             } else {
                 Err(SbiError::InvalidParam)
             };
-            // TODO: Update internal counter state on success.
-            result.map_err(|e| EcallError::Sbi(e))
+            result
+                .map(|n| {
+                    active_cpu.update_stopped_counters(
+                        counter_index_base,
+                        counter_index_mask,
+                        stop_flags,
+                    );
+                    n
+                })
+                .map_err(|e| EcallError::Sbi(e))
         }
 
         #[allow(clippy::redundant_closure)]
         fn configure_counters(
-            counter_index_base: u64,
+            counter_index: u64,
             counter_index_mask: u64,
             config_flags: PmuCounterConfigFlags,
             event_type: PmuEventType,
             event_data: u64,
-            _active_cpu: &mut VmCpu,
+            active_cpu: &mut VmCpu,
         ) -> EcallResult<u64> {
             use sbi::api::pmu;
             let pmu_info = PmuInfo::get();
-            let result = if let Some(info) = pmu_info.get_counter_info(counter_index_base) {
+            let result = if active_cpu.is_counter_configurable(counter_index, config_flags) && let Some(info) = pmu_info.get_counter_info(counter_index) {
                 let config_flags = config_flags.set_sinh().set_minh();
                 if info.is_hardware_counter() {
                     pmu::configure_matching_counters(
```

Comment:
If we configure the counter here, our attempt to configure the counter when the vCPU is run will fail. We really should only be configuring the counter when we swap in the vCPU.

That said, we need the counter index to be assigned at this point. So do we configure + unconfigure? Hm.